### ISSUE TITLE
incremental indexing fails with "attempting to add file with date matching deleted document" due to Git being too smart

### ISSUE THREAD

**Issue by vladak** [oracle/opengrok#4607]
When incrementally indexing the https://github.com/openssl/openssl/ repository recently, the indexer failed with:
```
2024-07-24 08:00:29.443+0200 FINEST t42 RepositoryWithHistoryTraversal.doCreateCache: storing history cache for revision range (4f619ca622b6c36626ddc9a04b0b8589d7802dc0, null)
...

2024-07-24 08:00:34.443+0200 FINEST t58 IndexDatabase.indexDownUsingHistory: collected sorted files: [/openssl/apps/fipsinstall.c, /openssl/doc/man1/openssl-fipsinstall.pod.in, /openssl/doc/man7/EVP_KDF-SS.pod, /openssl/doc/man7/EVP_KDF-SSHKDF.pod, /openssl/doc/man7/EVP_KDF-TLS13_KDF.pod, /openssl/doc/man7/EVP_KDF-TLS1_PRF.pod, /openssl/doc/man7
/EVP_KDF-X963.pod, /openssl/include/openssl/fips_names.h, /openssl/providers/common/include/prov/fipscommon.h, /openssl/providers/common/include/prov/securitycheck.h, /openssl/providers/common/securitycheck_default.c, /openssl/providers/common/securitycheck_fips.c, /openssl/providers/fips/fipsprov.c, /openssl/providers/implementations/kdfs/hkdf.
c, /openssl/providers/implementations/kdfs/sshkdf.c, /openssl/providers/implementations/kdfs/sskdf.c, /openssl/providers/implementations/kdfs/tls1_prf.c, /openssl/test/evp_kdf_test.c, /openssl/test/recipes/30-test_evp_data/evpkdf_hkdf.txt, /openssl/test/recipes/30-test_evp_data/evpkdf_ss.txt, /openssl/test/recipes/30-test_evp_data/evpkdf_ssh.txt
, /openssl/test/recipes/30-test_evp_data/evpkdf_tls12_prf.txt, /openssl/test/recipes/30-test_evp_data/evpkdf_tls13_kdf.txt, /openssl/test/recipes/30-test_evp_data/evpkdf_x963.txt, /openssl/test/recipes/30-test_evp_data/evppkey_kdf_hkdf.txt, /openssl/test/recipes/30-test_evp_data/evppkey_kdf_tls1_prf.txt, /openssl/test/recipes/80-test_cms.t, /ope
nssl/util/mk-fipsmodule-cnf.pl, /openssl/util/perl/OpenSSL/paramnames.pm]
...
2024-07-24 08:00:35.068+0200 FINE t58 DefaultIndexChangedListener.fileRemove: Remove: '/openssl/doc/man7/EVP_KDF-SS.pod'
2024-07-24 08:00:35.070+0200 FINER t58 DefaultIndexChangedListener.fileRemoved: Removed: '/openssl/doc/man7/EVP_KDF-SS.pod'
2024-07-24 08:00:35.109+0200 SEVERE t58 IndexDatabase.update: Failed with unexpected RuntimeException
org.opengrok.indexer.index.IndexDatabase$IndexerFault: attempting to add file '/openssl/doc/man7/EVP_KDF-SS.pod' with date matching deleted document: 20231107150430739
        at org.opengrok.indexer.index.IndexDatabase.processFileHistoryBased(IndexDatabase.java:1779)
        at org.opengrok.indexer.index.IndexDatabase.indexDownUsingHistory(IndexDatabase.java:985)
        at org.opengrok.indexer.index.IndexDatabase.getIndexDownArgs(IndexDatabase.java:946)
        at org.opengrok.indexer.index.IndexDatabase.update(IndexDatabase.java:755)
        at org.opengrok.indexer.index.Indexer.lambda$doIndexerExecution$59(Indexer.java:1208)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)   
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)                      
```
Indeed, the file had its last modified time stamp in the past:
```
$ stat /ws-local/openssl/doc/man7/EVP_KDF-SS.pod      
  File: /ws-local/openssl/doc/man7/EVP_KDF-SS.pod                               
  Size: 5686            Blocks: 7          IO Block: 6144   regular file        
Device: d800010031h/927713001521d       Inode: 232427131   Links: 1             
Access: (0644/-rw-r--r--)  Uid: (735504/wsmirror)   Gid: (   10/   staff)       
Access: 2023-11-07 16:04:30.738751219 +0100                                     
Modify: 2023-11-07 16:04:30.739240013 +0100                                     
Change: 2023-11-07 16:04:30.739240013 +0100                                     
 Birth: 2023-11-07 16:04:30.738751219 +0100        
```
hence the failure.

Hard-resetting the repository to `001b92d68d61250c88b355773142af31675ca0ab` (grandparent of `4f619ca622b6c36626ddc9a04b0b8589d7802dc0`) and then performing  `git pull -v --ff-only` the list of changed files produced by the command lacked `doc/man7/EVP_KDF-SS.pod `.

There were 2 changesets which modified the file: `6d47e819f2101f0219ddee67e855701e7bc3a716`, `14e46600c68ece74970462a60ad20703221747a1` however the changes there cancelled each other out and it seems Git is smart enough to filter the file out when performing pull across both changesets. The history based reindex does not consider file contents when constructing the list of files, though.

This is something which should be investigated for Mercurial as well.

**Comment by cesar-d2l** [comment]
Hey Vladimir. I'm hitting this issue in 1.13.17 (which GitHub tells me includes the f7dbff5 fix). Should I open up a new issue for it?

**Comment by vladak** [comment]
> Hey Vladimir. I'm hitting this issue in 1.13.17 (which GitHub tells me includes the f7dbff5 fix). Should I open up a new issue for it?

did you reindex from scratch after updating ?

**Comment by cesar-d2l** [comment]
> > Hey Vladimir. I'm hitting this issue in 1.13.17 (which GitHub tells me includes the [f7dbff5](https://github.com/oracle/opengrok/commit/f7dbff57fb9fdb17cd4032c69adafbbf2459d47d) fix). Should I open up a new issue for it?
> 
> did you reindex from scratch after updating ?

Yes (the way we have it deployed, each OpenGrok update gets it's down index)

**Comment by Shivareddy-Aluri** [comment]
Hi @vladak , we are still hitting this issue. i'm on 1.14.13.

Note that f7dbff5 ("History based reindex vs. git revert", #4609) is already included in 1.14.13 (it's an ancestor of the 1.14.13 tag), so this build has that fix — but the fault still reproduces. It looks like getDocument(file) returns null → the guard (in isStrictlyNewerThanDocument) hits the return true

Scenario (real example, file .java-version):

Between the last-indexed revision and current HEAD, two commits touch the file and cancel out:
- commit A: 17.0 → 11.0 (a revert)
- commit B: 11.0 → 17.0 (re-apply)

Net content is unchanged (still 17.0), so:
  - git diff <lastIndexedRev> <HEAD> does not list the file, and git does not rewrite it in the working tree → the file's mtime is unchanged (stale).
  - But both commits touched it, so history-based collection (FileCollector, via git log-style per-changeset accumulation) does include it in the reindex set.

OpenGrok then removes the existing document and re-adds the file, but since the mtime never changed, the new uid (path + mtime) is identical to the one it just deleted → IndexerFault. The file's mtime equals the
  timestamp in the error (20260719132438752), confirming the file on disk was never rewritten.

latest trace:
```
WARNING: cannot get document for ‘**/**/**/.java-version'
Aug 22, 2026 4:36:41 PM org.opengrok.indexer.index.IndexDatabase update
SEVERE: Failed with unexpected RuntimeException
org.opengrok.indexer.index.IndexDatabase$IndexerFault: attempting to add file ‘/repo/folder/file' with date matching deleted document: 20260719132438752
	at org.opengrok.indexer.index.IndexDatabase.processFileHistoryBased(IndexDatabase.java:1721)
	at org.opengrok.indexer.index.IndexDatabase.indexDownUsingHistory(IndexDatabase.java:930)
	at org.opengrok.indexer.index.IndexDatabase.getIndexDownArgs(IndexDatabase.java:845)
	at org.opengrok.indexer.index.IndexDatabase.update(IndexDatabase.java:657)
	at org.opengrok.indexer.index.IndexDatabase.lambda$updateAll$5(IndexDatabase.java:346)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

is there any plan to fix this? as this is stopping us from running incremental re-indexing with history enabled.
